File: backend/app/models/analysis.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class SessionStatus(str, enum.Enum):
    PENDING = "pending"
    ALIGNING = "aligning"
    ANALYZING = "analyzing"
    COMPLETE = "complete"
    FAILED = "failed"


class AnalysisCategory(str, enum.Enum):
    PITCH = "pitch"
    RHYTHM = "rhythm"
    TEMPO = "tempo"
    DYNAMICS = "dynamics"


class AnalysisSession(Base):
    __tablename__ = "analysis_sessions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), nullable=False)
    reference_recording_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("recordings.id"), nullable=False)
    student_recording_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("recordings.id"), nullable=False)
    status: Mapped[SessionStatus] = mapped_column(
        Enum(SessionStatus, values_callable=lambda enum_cls: [e.value for e in enum_cls]),
        default=SessionStatus.PENDING,
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user: Mapped["User"] = relationship(back_populates="analysis_sessions")
    results: Mapped[list["AnalysisResult"]] = relationship(back_populates="session")
    feedback_items: Mapped[list["Feedback"]] = relationship(back_populates="session")


class AnalysisResult(Base):
    __tablename__ = "analysis_results"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("analysis_sessions.id"), nullable=False)
    category: Mapped[AnalysisCategory] = mapped_column(
        Enum(AnalysisCategory, values_callable=lambda enum_cls: [e.value for e in enum_cls]),
        nullable=False,
    )
    data: Mapped[dict] = mapped_column(JSONB, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    session: Mapped["AnalysisSession"] = relationship(back_populates="results")

File: frontend/src/types/analysis.ts
export type AnalysisCategory = "pitch" | "rhythm" | "tempo" | "dynamics";
export type SessionStatus = "pending" | "aligning" | "analyzing" | "complete" | "failed";

// These shapes are deliberately kept in exact sync with the dicts built
// in backend/worker/tasks/analyze.py's results_payload — that's the one
// and only place this JSON is produced, so these types should change
// only if that function's output changes.

export interface FlaggedRegion {
  start: number;
  end: number;
}

export interface LabeledFlaggedRegion extends FlaggedRegion {
  label: string;
}

export interface PitchResultData {
  mean_absolute_cents_deviation: number;
  flagged_regions: FlaggedRegion[];
  points: { reference_time: number; cents_deviation: number }[];
}

export interface RhythmResultData {
  mean_absolute_timing_offset_seconds: number;
  unmatched_reference_onsets: number;
  unmatched_student_onsets: number;
  flagged_regions: FlaggedRegion[];
}

export interface TempoResultData {
  mean_tempo_ratio: number;
  reference_average_bpm: number | null;
  student_average_bpm: number | null;
  flagged_regions: LabeledFlaggedRegion[];
  points: { reference_time: number; local_tempo_ratio: number }[];
}

export interface DynamicsResultData {
  mean_absolute_loudness_difference_db: number;
  flagged_regions: LabeledFlaggedRegion[];
}

export interface AnalysisResult {
  id: string;
  session_id: string;
  category: AnalysisCategory;
  data: Record<string, unknown>;
  created_at: string;
}

export interface Feedback {
  id: string;
  session_id: string;
  category: AnalysisCategory;
  text: string;
  timestamp_reference: number | null;
  created_at: string;
}

export interface AnalysisSession {
  id: string;
  user_id: string;
  reference_recording_id: string;
  student_recording_id: string;
  status: SessionStatus;
  created_at: string;
  results: AnalysisResult[];
  feedback: Feedback[];
}
